ADD for serializable objects (#813)
pace-neutrons/Horace: documentation/add/09_Serializable_Interfaces.md

# Serialiable interfaces
Date: 2022-06-01

## Objectives and problems to solve.

> Current Horace SQW objects use its native proprietary (though an open source) binary file format to store Horace `sqw` objects and allow fast access to these objects. The problem is in accessing pixel information in fact the data, containing all data records from a weeks long inelastic scattering neutrons experiments. The main reason for introducing such format is that these data can not fit to memory of common computers (average size of the files would be *~50Gb* but *500Gb* files are not so uncommon), so we need to keep these data on a disk and provide efficient ways of accessing and processing them.

Author: hopefully rewritten clearler

> Current Horace SQW objects use its native proprietary (though an open source) binary file format to store Horace `sqw` objects and allow fast access to these objects. The problem is in accessing pixel information in fact the data, containing all data records from a weeks long inelastic scattering neutrons experiments. The main reason for introducing such format is that these data can not fit to memory of common computers (average size of the files would be *~50Gb* but *500Gb* files are not so uncommon), so we need to keep these data on a disk and provide efficient ways of accessing and processing them.

Author: hopefully clarified

> Current Horace SQW objects use its native proprietary (though an open source) binary file format to store Horace `sqw` objects and allow fast access to these objects. The problem is in accessing pixel information in fact the data, containing all data records from a weeks long inelastic scattering neutrons experiments. The main reason for introducing such format is that these data can not fit to memory of common computers (average size of the files would be *~50Gb* but *500Gb* files are not so uncommon), so we need to keep these data on a disk and provide efficient ways of accessing and processing them.

Author: I would use that but have already modified the chapter according to Chris requests. Hope it is clear enough.

> In addition to that, users often work with parts of the whole experiment data, containing the areas of the interests for the user. These data can often, though not always, fit into memory and users want to store their areas of interest for further usage. The data in memory are presented as Matlab `sqw` objects so users often use Matlab proprietary file format, which allows efficient binary store/restore operations for a Matlab classes.

Author: hopefully clarified

> In addition to that, users often work with parts of the whole experiment data, containing the areas of the interests for the user. These data can often, though not always, fit into memory and users want to store their areas of interest for further usage. The data in memory are presented as Matlab `sqw` objects so users often use Matlab proprietary file format, which allows efficient binary store/restore operations for a Matlab classes.

Author: ok, would take this.

> To satisfy users requests, we in fact support two independent binary file formats. Each format is best suited for subset of user needs.

Author: don't I introduced them in the previous two chapters?

> To satisfy users requests, we in fact support two independent binary file formats. Each format is best suited for subset of user needs.

Author: hopefully clarified

> To satisfy users requests, we in fact support two independent binary file formats. Each format is best suited for subset of user needs.

Author: we currently not support them -- the serializable interface would reduce current mess to `saveobj` in `serializable`

> The fact that the data are binary and proprietary causes some problems for users. To access native Horace file format, users need deep understanding of Horace, as the binary objects are reflection of sqw objects in memory and these objects have complex structure necessary for storing complex experimental data. Users who want to utilize smaller `sqw` objects stored in `.mat` files should also use Horace as restoring Matlab classes relies on Matlab knowing the definition of these classes.

Author: to us (Horace) of course -- there is GPL license

> The fact that the data are binary and proprietary causes some problems for users. To access native Horace file format, users need deep understanding of Horace, as the binary objects are reflection of sqw objects in memory and these objects have complex structure necessary for storing complex experimental data. Users who want to utilize smaller `sqw` objects stored in `.mat` files should also use Horace as restoring Matlab classes relies on Matlab knowing the definition of these classes.

Author: nobody examine the format -- users want data. Binary SQW is not a direct representation -- its bloody convoluted reflection. Splitting sentence makes sence.

> 1. To satisfy the request of accessing Horace data from third party applications, team have decided to change Horace file format from raw binary, to [HDF](https://www.hdfgroup.org/) file format, as this format is the industry standard for efficient storage and access to binary scientific data, accessible by number of third party applications unrelated to Matlab and Horace. The decision on making `.hdf` data [NeXus](https://www.nexusformat.org/) compatible is still pending (NeXus is an HDF-based standard format for storing the results of neutron scattering experiments).

Author: I've added sentence about difficulties of implementing NeXus compatibility, but omitted benefits as this is actually not the place to discuss it. Would be nice to implement NeXus but do we have enough resources for that -- probably not.

> 1. To satisfy the request of accessing Horace data from third party applications, team have decided to change Horace file format from raw binary, to [HDF](https://www.hdfgroup.org/) file format, as this format is the industry standard for efficient storage and access to binary scientific data, accessible by number of third party applications unrelated to Matlab and Horace. The decision on making `.hdf` data [NeXus](https://www.nexusformat.org/) compatible is still pending (NeXus is an HDF-based standard format for storing the results of neutron scattering experiments).

Author: mainly accepted

> 2. Current binary file format is relatively complex and related to current structure of `sqw` classes. To satisfy the project requests we are bringing substantial changes to `sqw` objects so the file format to store these objects should also change. To maintain consistent user experience we need to support the way of reading various previous versions of the `sqw` binary files and `sqw` objects, stored in proprietary `.mat` files.

Author: mainly accepted

3. Additional problem to resolve is the maintenance of two independent file formats. Any changes to `sqw` objects would request changes in two independent file writers which requests additional developer's efforts. It would be beneficial to avoid duplication and maintain only one file format common for both for large (partially fitting to memory) and small (fully fitting to memory) `sqw` objects, written to disk.

## Suggested solution -- serializable interface.

> To resolve issues 1-3 mentioned in the previous chapter, team decided to rely on the Matlabs standard mechanism of storing/restoring customized objects. If a Matlab object defines `saveobj/loadobj` methods, Matlab uses these methods to convert to/from binary format, to convert a Matlab object into a structure or to recover the object from the structure. The structure then is saved/loaded using Matlab proprietary file format. The responsibility of maintaining this format is then lies with Matlab.

Author: no, as python wraps Matlab

> To utilize the `saveobj/loadobj` Matlab functionality we have decided to make all Horace objects `serializable`. The custom `serializable` class defines `saveobj/loadobj` pair of methods and some additional methods, necessary to maintain class versioning (see below).

Author: hopefully resolved

To implement our *serializable* interface we have to make two assumptions about our objects:

1. The object has an empty constructor.

> 2. The object has a public interface which allows to define(assign) to an empty object any contents the non-empty object may have, defining in this way any valid non-empty object.

Author: both explanation look identical to me, but if you prefer second explanation, I do not mind.

If such assumptions are satisfied, we may define *serializable* objects, which need to overload only handful of class-specific methods, but would immediately have number of very useful generic features. The class diagram describing such object is presented on the **Fig 1**:

> <center>![Serialiable. Main Interface](../diagrams/serializable_main_interface.png)</center>

Author: I made the definition of the second loadobj more distinctive and discuss its purpose later in the text

**Fig 1:** Main interface and methods of **serializable** class.

> The parent class, presented on the **Fig 1**, allows easy construction and maintenance of standardized serializable objects by overloading the following abstract methods:

Author: accepted

**Table 1: Main abstact methods requesing overloading (definition)**

| Method | Description |
|-----|---------|
|  classVersion | returns the number, which describes current class version. E.g. 1 - for version 1|

> | saveableFields | returns list of property names, defining the class public interface |

Author: despite Matlab uses this definition, its not a good definition here. We made it `dependent` to apply validators at getters/setters but they are not dependent per meaning. In fact, we may define them as independent properties -- `serializable` will work.

As soon as these properties are defined, one can use other class methods to serialize/deserialize objects and convert objects to/from structure. The main features of the available methods are summarized in the Tables 2:
** Table 2 Main methods, available for serializable objects:**

| Method | Description |
|-----|---------|

> |  `struct = to_bare_struct(obj)` | returns structure, with the field names equal to the names returned by `saveableFields` method and values, defined by the values of these fields. If the values of fields are the serializable objects themselves, they are converted into correspondent structures recursively |

Reviewer: are THE serializable objects - delete THE
correspondent -> corresponding

| `obj = from_bare_struct(obj,data)` | Reverse of `to_bare_struct` method. Sets the values of the public properties of the serializable interface from the structure, obtained from `to_bare_struct` method |

> | `data = to_struct(obj)` | converts `obj` or array of `obj` into the `data` structure using `to_bare_struct` method and adds complimentary information about the class name and the size and shape of the `obj` array if relevant. |

Reviewer: complimentary->complementary (complImentary would be "wow, this is a really cute struct")
I would have said additional, but your call on that

| *Static:* | |
| `obj = from_struct(data)` | Reverse of `to_struct` method, recovering the object or array of objects from the `data`, generated by the `to_struct` method |

> Standard `loadobj/saveobj` and `serialize/deserialize` methods use the data obtained from the methods described in **Table 2** or the methods themselves to provide/use input for/from Matlab save/load methods or serialize/deserialize methods to save/load data to/from `.mat` files or transform the objects from/to linear arrays of bytes for communications over serial pipes. The `serialize/deserialize` routines use Herbert-defined `serialize/deserialize` methods over the structures obtained using `to_struct/from_struct` methods.

Reviewer: Not clear about the last sentence here. I think you want to say that either we use to_struct/from_struct and put/get a struct, or we use serialize/deserialize and get a serialization string. Iam not clear why one would want to use serialise/deserialise

Author: you would use `serialize/deserialize` to send object over MPI pipe

> Two remaining methods presented on the **Fig 1** namely static `loadobj` and `from_old_struct` methods need overloading if/when you want to support loading the versions of a class, different from the the current class version. By default, a standard `loadobj` method exists and this method expects to receive the structure, produced by `to_struct` method. `to_struct` method adds information about the class name, so generic `loadobj` extracts this information and recovers the class. If this information is missing, which may happen when you loading a class structure, generated before the object become `serializable`, you need to overload `loadobj` to provide the empty instance of the class, as described by the code snippet below:

Reviewer: versions of a class, different from the current class version
(1) remove comma after class
(2) different from -> earlier than
New paragraph after "from the current class version2.

So you're saying that if I save an object which is already serializable, then the custom loadobj here is not necessary to retrieve it. It could be worth saying how the code knows that this is the case and does  not call the custom loadobj.

Author: 2 -- different -- comparison goes on equality and its theoretically possible to have version higher then current. You may want to downgrade

Reviewer 2: How about that?

```Matlab
        methods(Static)
            function obj = loadobj(S)
                % boilerplate loadobj method, calling generic method of
                % saveable class. Put it as it is replacing the
                "ChildClass" name by the name of the class the loadobj is
                the method of.
                obj = ChildClass();
                obj = loadobj@serializable(S,obj);
            end
        end
```
By default, the method `from_old_struct` calls `from_bare_struct` method, so it does not need overloading if the structure of your class public interface have not changed.

> If it have changed, `from_old_struct` needs overloading with the code, converting the old information, stored in the previous class structure into the information, necessary to define new class.

Author: Well, many useful changes but I've modified the text to address Chris's concerns. And the suggested changes to row 54 are completely wrong. I have modified the text to hopefully make the changes clearer

## Interdependent properties problem.

> All serializable interface described above and used inside the code is build under assumption that you define class value by assigning the values of the properties one by one. 

Reviewer: I think we need more than this as from_old_struct has to consider a variety of older version. Some examples would be good. But it is probably not needed immediately

Author: This is ADD after all  -- ideas, not detailed instruction. I would refer to existing code for examples.

All public setters run checks of the validity of the values one wants to assign to the properties and throw exceptions if the input values are invalid. This assumption causes one problem with interdependent properties. For example, arrays `s`,`e`, and `npix` of a `dnd` or `sqw` object need to be the same shape and size arrays. If you are modifying, say, empty `dnd` object assigning `s`, `e` and `npix` arrays one after another, the first assignment (for `s`) would be always invalid, as the sizes of two other arrays are different from the first one as the initial object was containing empty data in arrays `e` and `npix`.

> There is number of ways, one can deal with this issue. After discussion, we have decided to select the following approach, summarized on the **Fig 2** describing the interface  used to validate interdependent properties. 

Author: accepted

<center>![serializable Validation Interface](../diagrams/Serializable_validation_interface.png)</center>

> **Fig 2:** Validation interface for **serializable** class.

Reviewer: Our normal naming convention is that A is a public property and A_ is its private implementation. Here check_combo_arg_ is a logical encapsulating the state that check_combo_arg looks at. I think the names should change to reflect the different function

Author: the idea was not to have public property `check_combo_arg` for protected property `check_combo_arg_`. But may be it makes sense to have one (hidden). Then the renaming is justified

 Any interdependent properties for a child of the *serializable* class, which has interdependent properties, have to overload `[ok,mess,obj] = check_combo_arg(obj)` method. This method verifies validity of interdependent properties. Each setter for interdependent properties runs this method within the following code snippet:
 ```Matlab
    function obj = set.an_interdependent_prop(obj,val)
        check_general_acceptance_of_input_throw_if_invalid(val);
        if obj.check_combo_arg_
            [ok,mess,obj] = check_combo_arg(obj,val)
            if ~ok
                error('HORACE:interdependent_properties_validation_error',message)
            end

>         end

Author: I thought about it and probably implement the throwing inside the check_combo arguments to avoid code duplication. For illustration purposes it is certainly better to leave the snippet as it is -- it illustrates code behavior better

    end
 ```

> Before the properties are set within the serializable class constructor one by one, the *serializable* code sets the protected property **check_combo_arg_** to false, so the check for each interdependent property is not occurring. The code then sets the property **check_combo_arg_** to true and runs the check method: *check_combo_arg* after all interdependent properties have been set, throwing if **check_combo_arg** returns **false**. As the result, user can not change interdependent properties in a wrong way and needs to use constructor or serializable interface to set defined class with interdependent properties.

Reviewer: I am not clear how check_combo_arg_ gets set true. If all the properties have not been set, how is the use of the class set invalid?

Author: its protected property (or private, does not metter), which can be set/unset within the class methods.

> Before the properties are set within the serializable class constructor one by one, the *serializable* code sets the protected property **check_combo_arg_** to false, so the check for each interdependent property is not occurring. The code then sets the property **check_combo_arg_** to true and runs the check method: *check_combo_arg* after all interdependent properties have been set, throwing if **check_combo_arg** returns **false**. As the result, user can not change interdependent properties in a wrong way and needs to use constructor or serializable interface to set defined class with interdependent properties.

Author: that's incorrect -- the object gets instantiated (with empty constructor) and `from_struct` defines its contents. I have elaborated more to make it clearer

> There is other way of setting interdependent properties assuring their validity, namely defining the methods, which allow setting all interdependent properties at once. E.g. setting `s`, `e`, `npix` array may be performed by introducing property `se_npix` with validates and sets all three properties together. This method may be more efficient from point of view of serializing data so will be used where it is justified and convenient.

Author: much clearer

> ## Standard serializable class constructor form.

Reviewer: Presumably this is an ordinary constructor and has no relation to construction on load. Worth making this clear

Author: I've hopefully reformatted header to make it cleared

As a *serializable* class has public interface, fully described by the list of properties, returned by *saveableFields* method, it makes sense to define standard constructor, which use this interface to define the contents of the serializable class. The constructor would have the form:
```Matlab

>     obj = class_name(positional_val1,positional_val2,positional_val3,... '-key1',val1,'-key2',val2,...);

Reviewer 2: According to ADR [#22](https://github.com/pace-neutrons/Horace/blob/master/documentation/adr/0022-use-keyword-args.md) keyword vals should not be prefixed with `-`, only `true`/`false` flags should.

Author: ok, easier to make it this way.

```
where *positional_val1, positional_val2...* are the values, one would set up to the public properties of the *serializable* interface in the order, these properties names are returned by *saveableFields* method. The `keys` then would be the names of these properties, followed by the appropriate values.

> To facilitate creation of such constructors, the *serializable* class has a method:

Reviewer 2: ```suggestion
To facilitate creation of such constructors, the `serializable` class has a method:
```

```Matlab
 [obj,remains] = set_positional_and_key_val_arguments(obj,...
                param_names_list,varargin);
```

where `param_names_list` is the list of the property names. Then, a standard constructor for a serializable object would have a form:

```Matlab
    function obj = an_serializable_obj(varargin)
    	if nargin == 0
            return;
        end
        prop_names = obj.saveableFields();
        obj = set_positional_and_key_val_arguments(obj,...
                param_names_list,varargin{:});
    end
```

## Generic eq operator.

As serializable classes have public interfaces with values fully defining the state of the object, to compare serializable objects is reasonable to define `eq` operator, which would work by comparing the values, retrieved from public properties provided by *saveableFields* method of the serializable interface.

> To avoid code duplication, such operator should be defined directly on *serializable* interface.

Reviewer: Can you do this? The actual eq would depend on the class definition

Reviewer 2: It could be overloaded on a class by class basis and invoke the serializable class definiton if useful.

Author: I can do that -- all serializable classes expose the interface for comparison so `eq` will use the interface. Nobody prevents you from overloading parental `eq` operator to modify comparison details but the standard `eq` is good for majority of cases.

> As serializable classes have public interfaces with values fully defining the state of the object, to compare serializable objects is reasonable to define `eq` operator, which would work by comparing the values, retrieved from public properties provided by *saveableFields* method of the serializable interface.
> 
> To avoid code duplication, such operator should be defined directly on *serializable* interface.

Reviewer 2: ```suggestion
As `serializable` classes have public interfaces which fully define the state of the object, it is reasonable to define an `eq` operator to compare `serializable` objects, which would work by comparing the values provided by `saveableFields` method of the `serializable` interface.

To avoid code duplication, this operator should be defined within the `serializable` class.
```

Author: `saveableFields` returns list of properties, so the values are the values of the properties and not "the values provided by `saveableFields` method of the `serializable` interface"
